File: cpp/297_SerializeAndDeserializeBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "";
        string serialized_str = to_string(root->val);
        vector<TreeNode*> stack = {root};
        
        // iterative dfs
        while (!stack.empty()) {
            TreeNode* curr = stack.back();
            stack.pop_back();
            
            if (curr) {
                // serialized string takes left and right of curr node
                serialized_str += curr->left ? " " + to_string(curr->left->val) : " N";
                serialized_str += curr->right ? " " + to_string(curr->right->val) : " N";
                stack.push_back(curr->right);
                stack.push_back(curr->left);
            }
        }
        return serialized_str;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data == "") return NULL;
        
        int i = 0; // deserialization pointer for data
        
        // get root val
        string num = "";
        while (data[i] != ' ')
            num += data[i++];
        i++;
        
        TreeNode* root = new TreeNode(stoi(num));
        vector<TreeNode*> stack = {root};
        
        while (i < data.length() && !stack.empty()) {
            TreeNode* curr = stack.back();
            stack.pop_back();
            
            // build left node
            string left = "";
            while (i < data.length() && data[i] != ' ')
                left += data[i++];
            ++i;
            
            // build right node
            string right = "";
            while (i < data.length() && data[i] != ' ')
                right += data[i++];
            ++i;
            
            if (right != "" && right != "N") {
                curr->right = new TreeNode(stoi(right));
                stack.push_back(curr->right);
            }
            if (left != "" && left != "N") {
                curr->left = new TreeNode(stoi(left));
                stack.push_back(curr->left);
            }
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: python/297_SerializeAndDeserializeBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string.
        """
        if not root:
            return ''
        
        serialized_str = f'{root.val}'
        stack = [root]
        while stack:
            curr = stack.pop()
            if curr:
                serialized_str += f' {curr.left.val}' if curr.left else ' N'
                serialized_str += f' {curr.right.val}' if curr.right else ' N'
                stack.append(curr.right)
                stack.append(curr.left)
        
        return serialized_str
                
        

    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree.
        """
        if not data:
            return None
        
        encoded_data = deque(data.split(' '))
        root = TreeNode(encoded_data.popleft())
        stack = [root]
    
        while stack:
            curr = stack.pop()
            left = encoded_data.popleft()
            right = encoded_data.popleft() 
            
            if right != 'N':
                curr.right = TreeNode(right)
                stack.append(curr.right)
            if left != 'N':
                curr.left = TreeNode(left)
                stack.append(curr.left)
        
        return root
            
            
        
        

# Your Codec object will be instantiated and called as such:
# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# tree = ser.serialize(root)
# ans = deser.deserialize(tree)
# return ans
